File: samples/typescript/discovery/agent_proxy.ts
import { ChatAnthropic } from "@langchain/anthropic";

type AgentFn = (task: string) => Promise<string>;

function createProxy(agentFn: AgentFn, maxRetries = 2): AgentFn {
  return async (task: string): Promise<string> => {
    let attempt = 0;
    while (attempt <= maxRetries) {
      console.log(`[Proxy] Attempt ${attempt + 1} for task: "${task}"`);
      const start = Date.now();
      try {
        const result = await agentFn(task);
        const elapsed = Date.now() - start;
        console.log(`[Proxy] Success in ${elapsed}ms`);
        return result;
      } catch (err) {
        const elapsed = Date.now() - start;
        console.error(`[Proxy] Attempt ${attempt + 1} failed after ${elapsed}ms:`, err);
        attempt++;
        if (attempt > maxRetries) {
          throw new Error(`[Proxy] All ${maxRetries + 1} attempts failed.`);
        }
        console.log(`[Proxy] Retrying...`);
      }
    }
    throw new Error("[Proxy] Unexpected exit from retry loop.");
  };
}

async function coreAgent(task: string): Promise<string> {
  const llm = new ChatAnthropic({ model: "claude-haiku-4-5-20251001" });
  const res = await llm.invoke(task);
  return res.content as string;
}

async function main() {
  const proxiedAgent = createProxy(coreAgent, 2);

  const task = "List three advantages of using an agent proxy pattern in distributed systems.";
  const result = await proxiedAgent(task);

  console.log(`\nResult:\n${result}`);
}

main().catch(console.error);

File: samples/python/discovery/agent_proxy.py
import time
from langchain_anthropic import ChatAnthropic
from typing import Callable

# --- Agent Proxy Pattern ---

llm = ChatAnthropic(model="claude-haiku-4-5-20251001")


def underlying_agent(task: str) -> str:
    return llm.invoke(f"Complete this task concisely: {task}").content


def agent_proxy(
    task: str,
    agent_fn: Callable[[str], str],
    max_retries: int = 2,
) -> str:
    print(f"[Proxy] Received task: {task}")
    print(f"[Proxy] Translating protocol: raw string -> agent call")

    attempt = 0
    last_error = None

    while attempt <= max_retries:
        try:
            start = time.time()
            print(f"[Proxy] Attempt {attempt + 1}: calling underlying agent...")
            result = agent_fn(task)
            elapsed = time.time() - start
            print(f"[Proxy] Agent responded in {elapsed:.2f}s")
            print(f"[Proxy] Logging result (first 80 chars): {result[:80]!r}")
            return result
        except Exception as e:
            last_error = e
            attempt += 1
            print(f"[Proxy] Error on attempt {attempt}: {e}. Retrying...")

    raise RuntimeError(
        f"[Proxy] All {max_retries + 1} attempts failed. Last error: {last_error}"
    )


def main():
    task = "List 3 benefits of using an agent proxy pattern in distributed systems."
    print("=" * 60)
    print("Agent Proxy Pattern Demo")
    print("=" * 60)

    result = agent_proxy(task, underlying_agent)

    print("\n[Final Result]")
    print(result)


if __name__ == "__main__":
    main()
